src/gasman.c

```diff
@@ -2266,50 +2263,44 @@ UInt CollectBags (
     for (i = 0; i < CollectFuncBags.nrBefore; ++i)
         CollectFuncBags.before[i]();
 
-    /* copy 'full' into a global variable, to avoid warning from GNU C     */
-    FullBags = full;
-
     /* do we want to make a full garbage collection?                       */
-again:
-    if ( FullBags ) {
-
-        /* then every bag is considered to be a young bag                  */
-        YoungBags = OldBags;
-        NrLiveBags = 0;
-        SizeLiveBags = 0;
-
-        /* empty the list of changed old bags                              */
-        while ( ChangedBags != 0 ) {
-            first = ChangedBags;
-            ChangedBags = LINK_BAG(first);
-            LINK_BAG(first) = first;
+    do {
+        if (full) {
+
+            // then every bag is considered to be a young bag
+            YoungBags = OldBags;
+            NrLiveBags = 0;
+            SizeLiveBags = 0;
+
+            // empty the list of changed old bags
+            while ( ChangedBags != 0 ) {
+                Bag first = ChangedBags;
+                ChangedBags = LINK_BAG(first);
+                LINK_BAG(first) = first;
+            }
+
+            // Also time to change the tag for dead children of weak pointer
+            // objects. After this collection, there can be no more weak pointer
+            // objects pointing to anything with OldWeakDeadBagMarker in it.
+            SWAP(Bag *, OldWeakDeadBagMarker, NewWeakDeadBagMarker);
         }
 
-        // Also time to change the tag for dead children of weak pointer
-        // objects. After this collection, there can be no more weak pointer
-        // objects pointing to anything with OldWeakDeadBagMarker in it.
-        SWAP(Bag *, OldWeakDeadBagMarker, NewWeakDeadBagMarker);
-    }
+        // information at the beginning of garbage collections
+        SyMsgsBags(full, 0, 0);
 
-    /* information at the beginning of garbage collections                 */
-    SyMsgsBags(FullBags, 0, 0);
+        // mark phase
+        nrBags = CollectBags_Mark(full);
 
-    /* * * * * * * * * * * * * * *  mark phase * * * * * * * * * * * * * * */
-    nrBags = CollectBags_Mark();
+        // sweep phase
+        nrBags += CollectBags_Sweep(full);
 
-    /* * * * * * * * * * * * * * * sweep phase * * * * * * * * * * * * * * */
-    nrBags += CollectBags_Sweep();
+        // check phase
+        done = CollectBags_Check(size, full, nrBags);
 
-    /* * * * * * * * * * * * * * * check phase * * * * * * * * * * * * * * */
-    done = CollectBags_Check(size, nrBags);
-    if (done == 2)
```

Comment:
(Why) have you discarded this case?

Reply:
Oops, good catch, that was an accident. Well, the plan *is* to remove that, but then also the final `return` needs to be changed. But that latter change is missing here; and in any case, this change of how `done == 2` is handled should not be sneaked in silently like this as part of refactoring. I've now put it into its own commit, properly and with a commit message/

Note that this change will conflict with PR #3793, though resolving the conflict ought to be trivial.

Reply:
I've now moved this into its own PR #3812